GenotypeConcordance can output an annotated VCF.

nh13 opened the pull request:
I find it useful to manually inspect individual variant calls based on their concordance state using this tool.  I also find it useful for filtering for variants with a particular concordance state (ex. FP).

It should also be possible (not implemented here) to re-compute the genotype concordance metrics using the new output.  This would be useful if you want to remove variants based on some criteria (ex. low-frequency, info/format tag/value) and re-compute the metrics.  Of course, one could always filter the original input VCF that is given to `GenotypeConcordance`.

(@yfarjoun `||` @tfenne `||` @droazen) would you mind taking a look?

tfenne commented:
@nh13 This is fantastic and I'm excited to have this available.  I can probably review over the weekend.  Would be good for someone else to take a peek too probably though?

nh13 commented:
@droazen or @yfarjoun would you also care to take a look?  My two concerns wit this PR are (1) there is not a really good way to merge VCF headers, and (2) some of the methods I use to re-compute the concordance state for output to the VCF could be changed so that I am not re-computing.

yfarjoun commented:
This is a very nice thing to have @kbergin you would find use for this quite soon I think...care to review it?

nh13 commented:
Just as a head's up, the second commit caused me to refactor a lot of code, since I hit the case where the reference allele for two different variant contexts at the same side were different.  Therefore, I needed to normalize them, and that cause a domino effect in terms of needed to get access to certain internals, such as the normalized alleles for the truth and call variant contexts.  I also have increased my profound hate for the vcf code in htsjdk.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
 
+            // write out an annotated VCF
+            if (writer != null) {
+                final VariantContextBuilder builder;
```
I think this could move down into your `if` now?

nh13 replied:
Done

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
+                // Get the alleles for each genotype
+                final Alleles alleles = normalizeAlleles(truthContext, TRUTH_SAMPLE, callContext, CALL_SAMPLE);
+
+                // Ignore if both are one of mixed, filtered, or missing.
```
Is this comment still accurate?  I.e. alleles will be empty if both genotypes are mixed/filtered/missing?  If so, maybe reword to make that clear?

nh13 replied:
It is, will do.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
+
+                // Ignore if both are one of mixed, filtered, or missing.
+                if (!alleles.allAlleles.isEmpty()) {
+                    final List<Allele> allAlleles   = alleles.asAlleles();
```
`asAlleles` -> `asList`

nh13 replied:
Done.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
+                    }
+                    else {
+                        final GenotypeBuilder genotypeBuilder = new GenotypeBuilder(outputTruthSampleName);
+                        genotypeBuilder.alleles(Arrays.asList(Allele.NO_CALL, Allele.NO_CALL));
```
I think here, if `MISSING_SITES_HOM_REF == true` then it would be more helpful to produce an output genotype with ref/ref instead.

nh13 replied:
Done.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
+                        genotypeBuilder.attribute(VCFConstants.GENOTYPE_KEY, MISSING_VALUE_v4);
+                        genotypes.add(genotypeBuilder.make());
+                    }
+                    if (callContext != null && !callAlleles.isEmpty()) {
```
This whole `if` block is repetitious.  Could be extracted to a method that takes in the sample name, genotype and set of alleles...

nh13 replied:
Done.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
+
+                    // set the concordance state attribute
+                    final TruthAndCallStates state = GenotypeConcordance.determineState(truthContext, TRUTH_SAMPLE, callContext, CALL_SAMPLE, MIN_GQ, MIN_DP);
+                    final ContingencyState[] stateArray = scheme.getConcordanceStateArray(state.truthState, state.callState);
```
I think it would be nicer to output something other than the contingency values here because I think those are a bit confusing until you understand their purpose.  I.e. seeing something labelled `TP,FP` is a little odd.  But I don't think there's really much else that could be emitted instead that's ready to go.

nh13 replied:
I agree the output could be improved, lets think about it and add it in later.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
+        public final String callAllele1;
+        public final String callAllele2;
+
+
```
Extra newline.

nh13 replied:
Just making sure you were checking ;)

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
-        // TODO: what about getPloidy()
 
-        Genotype truthGenotype = null, callGenotype = null;
+        public List<Allele> asAlleles() {
```
`asList`

nh13 replied:
Done.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
+                return Collections.emptyList();
+            }
+            else {
+                final List<Allele> alleles = new ArrayList<>(4);
```
Either remove `4` or change to `this.allAlleles.size()` please.

nh13 replied:
Done.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
+            else {
+                final List<Allele> alleles = new ArrayList<>(4);
+                int idx = 0;
+                while (idx < this.allAlleles.size()) {
```
why not just a regular for loop?

nh13 replied:
Why indeed.  Fixed.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
-            // Note.  Genotype.isMixed means that it is called on one chromosome and NOT on the other
-            else if ((truthGenotype.isMixed())) truthState = TruthState.NO_CALL;
+        public List<Allele> truthAlleles() {
+            if (allAlleles.isEmpty() || this.truthAllele1 == null) {
```
It feels a little odd that in the constructor you check `truthAllele2` for null independently, but here you seem to assume it must be non-null if `truthAllele1` is.

nh13 replied:
I'll add a check in the constructor.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
-                // Note.  Genotype.isMixed means that it is called on one chromosome and NOT on the other
-            else if ((callGenotype.isMixed())) callState = CallState.NO_CALL;
+        public List<Allele> callAlleles() {
+            if (allAlleles.isEmpty() || this.callAllele1 == null) {
```
Same here about the null-checking.

nh13 replied:
Fixed.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -457,70 +584,89 @@ public static void outputDetailMetricsFile(final VariantContext.Type variantType
+
+        final Genotype truthGenotype, callGenotype;
+
+        if (truthContext == null || truthContext.isMixed() || truthContext.isFiltered()) truthGenotype = null;
```
I'm not sure if it's done elsewhere, but before calling `isMixed()` would it make sense to produce sub-contexts that are just for the sample of interest?  I.e. we don't much care if a second sample has an allele that makes the context mixed.

nh13 replied:
The `PairedVariantSubContextIterator` creates a single-sample sub-context.  Yes, it's that confusing.

tfenne commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -546,19 +692,15 @@ final public static TruthAndCallStates determineState(final VariantContext truth
-                    truthAllele2 = truthGenotype.getAllele(1).getBaseString() + suffix;
-                }
+                truthAllele1 = truthGenotype.getAllele(0).getBaseString() + suffix;
+                truthAllele2 = truthGenotype.getAllele(1).getBaseString() + suffix;
```
Does this work properly on indels that are not repetitive?  I.e. shouldn't the `suffix` be spliced in between the shorter ref string and the rest of the allele?  My brain hurts thinking about this.

nh13 replied:
I think the current implementation works for a deletion, but not an insertion.  It should be added in-before the rest of the allele, which for deletions is empty, but for an insertion could be many bases.  I'll see if I can fix it.  Hrmn.

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -109,6 +108,13 @@
+            "" +
+            "<h4>VCF Output:</h4>" +
+            "<ul>" +
+            "<li>The concordance state will be stored in the \"ST\" tag in the INFO field.</li>" +
```
Will use `CONC_ST`

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -109,6 +108,13 @@
+            "<h4>VCF Output:</h4>" +
+            "<ul>" +
+            "<li>The concordance state will be stored in the \"ST\" tag in the INFO field.</li>" +
+            "<li>If the truth and call sample names match, the call will have the \"_CALL\" suffix and the truth will have the " +
```
Good idea, that makes it much simpler.

kbergin replied:
👍 that thought

nh13 replied:
@kbergin it's in a more recent commit, can you work off of that for your review?

kbergin replied:
Oh I did, I was just reading through what was commented on before your more recent changes out of curiousity, and wanted to "like" the change lol

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -118,9 +124,12 @@
     public File OUTPUT;
 
+    @Option(doc = "Output VCF annotated with concordance information.", optional = true)
+    public File OUTPUT_VCF = null;
```
Will do.

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -118,9 +124,12 @@
+    public File OUTPUT_VCF = null;
+
     @Option(shortName = "TS", doc="The name of the truth sample within the truth VCF")
     public String TRUTH_SAMPLE;
```
Not hard.

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -163,6 +172,9 @@
     protected GenotypeConcordanceCounts indelCounter;
     public GenotypeConcordanceCounts getIndelCounter() { return indelCounter; }
 
+    public static final String ContingencyStateTag = "ST";
```
+1 scala

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -163,6 +172,9 @@
     public GenotypeConcordanceCounts getIndelCounter() { return indelCounter; }
 
+    public static final String ContingencyStateTag = "ST";
+    public static final VCFHeaderLine ContingencyStateHeaderLine = new VCFInfoHeaderLine(ContingencyStateTag, VCFHeaderLineCount.UNBOUNDED, VCFHeaderLineType.String, "The genotype concordance contingency state(s)");
```
Done.

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -163,6 +172,9 @@
+    public static final String ContingencyStateTag = "ST";
+    public static final VCFHeaderLine ContingencyStateHeaderLine = new VCFInfoHeaderLine(ContingencyStateTag, VCFHeaderLineCount.UNBOUNDED, VCFHeaderLineType.String, "The genotype concordance contingency state(s)");
+
     // TODO: add optimization if the samples are in the same file
```
I don't have a need for it so I will defer to another PR.

tfenne replied:
I was suggesting deleting the three comment lines - not implementing them!

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -260,6 +276,30 @@ private boolean indexExists(final File vcf) {
         // Verify that both VCFs have the same Sequence Dictionary
         SequenceUtil.assertSequenceDictionariesEqual(truthReader.getFileHeader().getSequenceDictionary(), callReader.getFileHeader().getSequenceDictionary());
 
+        VariantContextWriter writer = null;
```
Done.

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -301,6 +341,89 @@ private boolean indexExists(final File vcf) {
             }
 
+            // write out an annotated VCF
+            if (writer != null) {
```
Done.

nh13 commented:
@tfenne, thanks for the comments, I implemented almost all of what you suggested.  @kbergin if possible, could you take a look?

kbergin commented:
@nh13 I have looked through this today and I think it all makes sense, but I'd like to look at it again on Monday when my brain has processed it more, it's a big change in code that always turns my head around! 

In the mean time, GenotypeConcordance tests are failing

nh13 commented:
@kbergin thanks for taking a look.  I have modified the tests too many times after folks' feedback, and it's painful to do, so I am going to wait until you have reviewed before I fix them up one last time (the output files differ, which is expected).  Thanks!

kbergin commented:
@nh13 That makes a lot of sense!!!!

kbergin commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -341,9 +375,119 @@ private boolean indexExists(final File vcf) {
+                                final List<Allele> allAlleles,
+                                final List<Allele> ctxAlleles,
+                                final boolean missingSitesHomRef) {
+        if (ctx != null && !ctxAlleles.isEmpty()) {
```
A comment explaining this function a little might be useful for future readers.

nh13 replied:
Fixed.

nh13 commented:
@kbergin just checking in to see if you have finished your review.

kbergin commented:
@nh13 Sorry, I was trying to follow the new github review structure, it seems now you 'start a review' and all the comments are pending until you mark it as a 'finished review' and then it posts it. So yup, all set! Can you see my couple comments from the other day?

nh13 commented on src/main/java/picard/vcf/GenotypeConcordance.java:
```diff
@@ -341,9 +375,119 @@ private boolean indexExists(final File vcf) {
+        // There will be no alleles if both genotypes are one of mixed, filtered, or missing.  Do not output any
+        // variant context in this case.
+        if (!alleles.allAlleles.isEmpty()) {
+
```
Fixed

nh13 commented:
@tfenne and @kbergin  I have fixed up the commits so I think it is ready for merge after tests pass, except....

...@yfarjoun `ByIntervalListVariantContextIterator` could return a variant multiple times if the variant overlaps multiple intervals.  I fixed that, and added tests (there were none).  Care to take a look at the first commit?   If you give me 👍 I can merge when tests pass.

kbergin commented:
Yay! 👍

yfarjoun commented on src/main/java/picard/vcf/ByIntervalListVariantContextIterator.java:
```diff
@@ -20,33 +23,58 @@
         throw new UnsupportedOperationException();
     }
 }
+
```
two extra newlines?

nh13 commented on src/main/java/picard/vcf/ByIntervalListVariantContextIterator.java:
```diff
@@ -20,33 +23,58 @@
+    private boolean overlapsInterval(final VariantContext ctx, final Interval interval) {
+        if (!ctx.getContig().equals(interval.getContig())) return false;
+        else if (CoordMath.overlaps(ctx.getStart(), ctx.getEnd(), interval.getStart(), interval.getEnd())) return true;
+        else if (CoordMath.encloses(ctx.getStart(), ctx.getEnd(), interval.getStart(), interval.getEnd())) return true;
```
Done.

nh13 commented:
Waiting for tests to pass.
